src/cata_utility.cpp

```diff
@@ -46,15 +46,16 @@ bool isBetween( int test, int down, int up )
 
 bool lcmatch( const std::string &str, const std::string &qry )
 {
-    std::string needle;
-    needle.reserve( qry.size() );
-    std::transform( qry.begin(), qry.end(), std::back_inserter( needle ), tolower );
+    std::locale::global( std::locale( "ru_RU.utf8" ) );
+    auto &f = std::use_facet<std::ctype<wchar_t>>( std::locale() );
```

Comment:
Not working:
```
src/cata_utility.cpp: In function ‘bool lcmatch(const string&, const string&)’:
src/cata_utility.cpp:60:16: error: invalid conversion from ‘const value_type* {aka const char*}’ to ‘std::ctype<char>::char_type* {aka char*}’ [-fpermissive]
```
Isn't wchar_t what we need here?

Type | Description
-- | --
std::ctype<char> | provides narrow character equivalents of the minimal "C" locale  classification. This specialization uses table lookup for character  classification
std::ctype<wchar_t> | provides wide character classification appropriate to the native character set

From [std::ctype](https://en.cppreference.com/w/cpp/locale/ctype).

Reply:
Please read the error messages: you are trying to use pointer to a `const char` in a place where a pointer to a (non-const) `char` is required (`tolower` changes the content of the string). Your code attempts to change `str` (which happens to be const, see function signature).

So to "fix" this, you make a copy (and convert it to `wstring` for no reason) and now you have  a non-const `wstring` and can use that.

Alternative suggestion: change the function signature to take a (non-const) value (`bool lcmatch( std::string str, std::string qry )`) and use that directly.

Reply:
Did I understand correctly that I would need to do something like this in all occurrences of `lcmatch`?
```diff
     switch( flag ) {
         case 'c'://category
             return [filter]( const item & i ) {
-                return lcmatch( i.get_category().name(), filter );
+                return lcmatch( utf8_to_wstr( i.get_category().name() ), utf8_to_wstr( filter ) );
             };
```

Reply:
If there are more than one or two call sites then it seems unhelpful to change them all thus.  It would make more sense to have two overloads of `lcmatch`: one for `string` and one for `wstring`.  One of those can be a simple wrapper function that calls the other with appropriate conversion.